crates/egui/src/memory.rs

```diff
@@ -322,31 +355,121 @@ impl Focus {
+                && !self.is_focus_locked
+            {
                 self.id_next_frame = self.last_interested; // frame-delay so gained_focus works
-                self.pressed_shift_tab = false;
+                self.reset_focus();
             }
-        } else if self.pressed_tab && self.id.is_none() && !self.give_to_next {
+        } else if matches!(self.focus_direction, FocusDirection::Next)
+            && self.id.is_none()
+            && !self.give_to_next
+        {
             // nothing has focus and the user pressed tab - give focus to the first widgets that wants it:
             self.id = Some(id);
-            self.pressed_tab = false;
+            self.reset_focus();
         }
 
         self.last_interested = Some(id);
     }
+
+    pub fn reset_focus(&mut self) {
+        self.focus_direction = FocusDirection::None;
+    }
+
+    pub fn find_widget_in_direction(&mut self, new_rects: &IdMap<Rect>) -> Option<Id> {
+        let Some(focus_id) = self.id else {return None;};
+
+        // Update cache with new rects
+        self.focus_widgets_cache.retain(|id, old_rect| {
+            if let Some(new_rect) = new_rects.get(id) {
+                *old_rect = *new_rect;
+                true // Keep the item
+            } else {
+                false // Remove the item
+            }
+        });
+
+        let current_focus_widget_rect = self.focus_widgets_cache.get(&focus_id).unwrap();
+        let current_widget_pos = current_focus_widget_rect.left_center();
+
+        let mut focus_candidates: Vec<(&Id, f32, f32)> =
+            Vec::with_capacity(self.focus_widgets_cache.len());
+
+        for (widget_id, widget_rect) in &mut self.focus_widgets_cache {
+            if Some(*widget_id) == self.id {
+                continue;
+            }
+
+            let candidate_widget_pos = widget_rect.left_center();
+
+            let current_to_candidate = candidate_widget_pos - current_widget_pos;
+
+            // Early out if widget is not aligned with the focus direction
+            let aligns_with_focus_direction = match self.focus_direction {
+                FocusDirection::Up => current_to_candidate.y < 0.0,
+                FocusDirection::Right => current_to_candidate.x > 0.0,
+                FocusDirection::Down => current_to_candidate.y > 0.0,
+                FocusDirection::Left => current_to_candidate.x < 0.0,
+                _ => false,
+            };
+
+            if !aligns_with_focus_direction {
+                continue;
+            }
+
+            // Check if the widget has the right dot product and length.
+            let focus_direction = match self.focus_direction {
+                FocusDirection::Up => Vec2::UP,
+                FocusDirection::Right => Vec2::RIGHT,
+                FocusDirection::Down => Vec2::DOWN,
+                FocusDirection::Left => Vec2::LEFT,
+                _ => {
+                    continue;
+                }
+            };
```

Comment:
If `current_to_candidate` is perpendicular to `focus_direction` it is 0. If `current_to_candidate` is parallel to `focus_direction` it is 0. `aligns_with_focus_direction` exists to prevent picking the perpendicular case. I do not see how this change solves this problem. Can you elaborate?

Reply:
Yes your right!